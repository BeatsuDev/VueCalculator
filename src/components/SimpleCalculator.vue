<script setup lang="ts">
import { useCalculationHistoryStore } from '@/stores/calculationHistory';
import { ref } from 'vue';

const display = ref<HTMLElement | null>(null);

var lastClickedWasSum = false;

function addSymbol(event: MouseEvent) {
  if (!event.target) return;
  if (!(event.target instanceof HTMLElement)) return;
  // Reset the display when a button is pressed IF:
  // 1. The display is currently showing an error
  // 2. The last button pressed was a sum button AND the current button is a number
  if (display.value!.innerHTML === 'Error' || (lastClickedWasSum && /^\d+$/.test(event.target.innerHTML))) {
    display.value!.innerHTML = '';
  }
  lastClickedWasSum = false;
  display.value!.innerHTML += event.target.innerHTML;
}

function calculate() {
  // Replace ^ with **
  let unchangedTask = display.value!.innerHTML;
  let toCalculate = display.value!.innerHTML.replace('^', '**');
  let answer;
  
  // If there are only numbers, don't try to calculate
  if (/^\d+$/.test(toCalculate)) {
    return;
  }
  try {
    answer = eval(toCalculate);
  } catch {
    display.value!.innerHTML = 'Error';
    return;
  }

  // Check if answer is valid
  if (isNaN(answer) || !isFinite(answer)) {
    display.value!.innerHTML = 'Error';
    return;
  }

  // Set the display to show the answer
  display.value!.innerHTML = answer;
  lastClickedWasSum = true;

  // Add to history
  useCalculationHistoryStore().addCalculation({
    task: unchangedTask,
    answer: answer.toString(),
  });
}

function clear() {
  display.value!.innerHTML = '';
}
</script>

<template>
  <main>
    <div id="display" ref="display">{{ display?.innerHTML }}</div>
    <div id="buttons">
      <div @click="addSymbol" class="button">(</div>
      <div @click="addSymbol" class="button">)</div>
      <div @click="addSymbol" class="button">^</div>
      <div @click="addSymbol" class="button">*</div>
      <div @click="addSymbol" class="button number">7</div>
      <div @click="addSymbol" class="button number">8</div>
      <div @click="addSymbol" class="button number">9</div>
      <div @click="addSymbol" class="button">/</div>
      <div @click="addSymbol" class="button number">4</div>
      <div @click="addSymbol" class="button number">5</div>
      <div @click="addSymbol" class="button number">6</div>
      <div @click="addSymbol" class="button">+</div>
      <div @click="addSymbol" class="button number">1</div>
      <div @click="addSymbol" class="button number">2</div>
      <div @click="addSymbol" class="button number">3</div>
      <div @click="addSymbol" class="button">-</div>
      <div @click="clear" class="button">cls</div>
      <div @click="addSymbol" class="button number">0</div>
      <div @click="calculate" class="button" id="ans-btn">Ans</div>
    </div>
  </main>
</template>

<style scoped>
main {
  background-color: #373652;
  border-radius: 10px;
  overflow: auto;
  border: 5px solid black;
  box-shadow: rgba(10, 10, 10, 0.8) 15px 15px 10px;
  resize: both;
  min-height: 320px;
  min-width: 320px;
}

#display {
  background-color: #bbb;
  color: var(--vt-c-white);
  border-radius: 10px;
  font-size: 25px;
  line-height: calc(40px - 2 * 5px);
  height: 40px;
  width: 95%;
  padding: 5px;
  margin: 10px auto;
  overflow-x: auto;
  overflow-y: hidden;
  white-space: nowrap;
}

#buttons {
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  grid-template-rows: repeat(5, 1fr);
  grid-gap: 2px;
  height: calc(100% - 40px - 10px - 10px);
}

.button {
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 5px;
  font-size: 1.5rem;
  padding: 5px;
  transition-duration: 100ms;
  user-select: none;
}

.button:hover {
  background-color: rgba(200, 200, 200, 0.5);
  cursor: pointer;
}

.number {
  background-color: #43426b;
}

#ans-btn {
  background-color: darkorange;
  grid-column: 3 / 5;
}

#ans-btn:hover {
  background-color: orange;
  color: white;
}
</style>