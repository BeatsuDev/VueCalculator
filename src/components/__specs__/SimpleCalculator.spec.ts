import { expect, describe, it, beforeEach } from "vitest";
import { DOMWrapper, mount, VueWrapper } from "@vue/test-utils";
import SimpleCalculator from "../SimpleCalculator.vue";

describe("SimpleCalculator", () => {
    let wrapper: VueWrapper;
    let number3: DOMWrapper<Element>;
    let ans: DOMWrapper<Element>;
    let display: DOMWrapper<Element>;

    beforeEach(() => {
        wrapper = mount(SimpleCalculator);

        number3 = wrapper.find("#buttons > div:nth-child(15)");
        ans = wrapper.find("#buttons > div:nth-child(19)");
        display = wrapper.find("#display");
    });

    it("should render", () => {
        const wrapper = mount(SimpleCalculator);
        expect(wrapper.html()).toMatchSnapshot();
    });

    it("clicking a number should run addSymbol() method", async () => {
        await number3.trigger("click");
        expect(wrapper.find("#display").text()).toBe("3");
    });

    it("should add two numbers", async () => {
        const plus = wrapper.find("#buttons > div:nth-child(12)");

        await number3.trigger("click")
        await plus.trigger("click");
        await number3.trigger("click");
        await ans.trigger("click");

        expect(display.text()).toBe("6");
    });
    
    it("should subtract two numbers", async () => {
        const minus = wrapper.find("#buttons > div:nth-child(16)");
        
        await number3.trigger("click");
        await minus.trigger("click");
        await number3.trigger("click");
        await ans.trigger("click");

        expect(wrapper.find("#display").text()).toBe("0");
    });

    it("should multiply two numbers", async () => {
        const multiply = wrapper.find("#buttons > div:nth-child(4)");
        
        await number3.trigger("click");
        await multiply.trigger("click");
        await number3.trigger("click");
        await ans.trigger("click");

        expect(wrapper.find("#display").text()).toBe("9");
    });

    it("should divide two numbers", async () => {
        const divide = wrapper.find("#buttons > div:nth-child(8)");
        
        await number3.trigger("click");
        await divide.trigger("click");
        await number3.trigger("click");
        await ans.trigger("click");

        expect(wrapper.find("#display").text()).toBe("1");
    });

    it("should clear the display", async () => {
        const clear = wrapper.find("#buttons > div:nth-child(17)");
        
        await number3.trigger("click");
        await clear.trigger("click");

        expect(wrapper.find("#display").text()).toEqual("");
    });

    it("should show error on divide by zero", async () => {
        const divide = wrapper.find("#buttons > div:nth-child(8)");
        
        await number3.trigger("click");
        await divide.trigger("click");
        await ans.trigger("click");

        expect(wrapper.find("#display").text()).toBe("Error");
    });

    it("should show error on invalid operation", async () => {
        const plus = wrapper.find("#buttons > div:nth-child(12)");
        await plus.trigger("click");
        await plus.trigger("click");
        await number3.trigger("click");
        await ans.trigger("click");

        expect(wrapper.find("#display").text()).toBe("Error");
    });
});