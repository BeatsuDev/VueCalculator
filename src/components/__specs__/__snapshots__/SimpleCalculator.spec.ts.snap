// Vitest Snapshot v1

exports[`SimpleCalculator > should render 1`] = `
"<main data-v-ae2a742a=\\"\\">
  <div data-v-ae2a742a=\\"\\" id=\\"display\\"></div>
  <div data-v-ae2a742a=\\"\\" id=\\"buttons\\">
    <div data-v-ae2a742a=\\"\\" class=\\"button\\">(</div>
    <div data-v-ae2a742a=\\"\\" class=\\"button\\">)</div>
    <div data-v-ae2a742a=\\"\\" class=\\"button\\">^</div>
    <div data-v-ae2a742a=\\"\\" class=\\"button\\">*</div>
    <div data-v-ae2a742a=\\"\\" class=\\"button number\\">7</div>
    <div data-v-ae2a742a=\\"\\" class=\\"button number\\">8</div>
    <div data-v-ae2a742a=\\"\\" class=\\"button number\\">9</div>
    <div data-v-ae2a742a=\\"\\" class=\\"button\\">/</div>
    <div data-v-ae2a742a=\\"\\" class=\\"button number\\">4</div>
    <div data-v-ae2a742a=\\"\\" class=\\"button number\\">5</div>
    <div data-v-ae2a742a=\\"\\" class=\\"button number\\">6</div>
    <div data-v-ae2a742a=\\"\\" class=\\"button\\">+</div>
    <div data-v-ae2a742a=\\"\\" class=\\"button number\\">1</div>
    <div data-v-ae2a742a=\\"\\" class=\\"button number\\">2</div>
    <div data-v-ae2a742a=\\"\\" class=\\"button number\\">3</div>
    <div data-v-ae2a742a=\\"\\" class=\\"button\\">-</div>
    <div data-v-ae2a742a=\\"\\" class=\\"button\\">cls</div>
    <div data-v-ae2a742a=\\"\\" class=\\"button number\\">0</div>
    <div data-v-ae2a742a=\\"\\" class=\\"button\\" id=\\"ans-btn\\">Ans</div>
  </div>
</main>"
`;
